src/hyperspectral/hyperspectral/target/rpca.py

```diff
@@ -0,0 +1,123 @@
+import math
+
+import numpy as np
+import scipy
+from tqdm.autonotebook import tqdm
+
+
+def mad(M):
+    return np.median(np.abs(M - np.median(M, axis=0)), axis=0)
+
+
+def rpca_grid(data, max_dim=None, n_c=5, n_g=11, S=mad, sufficient=None):
+    assert len(data.shape)==2, "Provide data as matrix of column vectors"
+    if n_g % 2 == 0:
+        # The search seems to fail when n_g is even
+        n_g = n_g + 1
+
+    p, n = data.shape
+    max_dim = max_dim if max_dim else p
+
+    # Order samples so that S(e(i)) >= S(e(j)) when i < j
+    row_scores = np.array([S(data[i,:]) for i in range(p)])
+    var_order = [y[1] for y in sorted([(x[1], x[0]) for x in enumerate(row_scores)], reverse=True)]
+    rev_order = [y[1] for y in sorted([(x[1], x[0]) for x in enumerate(var_order)])]
+    Xt = data[var_order,:].transpose()
+
+    def e(j):
+        x = np.zeros((p,1))
+        x[j] = 1.0
+        return x
+
+    A = np.zeros((p, max_dim))
+    for k in tqdm(range(max_dim), desc="Vector"):
+        for i in tqdm(range(n_c), leave=False, desc="Iteration"):
+            prev = float('nan')
+            with tqdm(total=p, bar_format="{percentage:3.0f}%|{bar}| {n_fmt}/{total_fmt} {postfix[0]} {postfix[1][value]:>6.4g} [{elapsed}<{remaining}]",
+              postfix=["Score", dict(value=float('nan'))], leave=False) as t:
+                for j in range(p):
+                    if i==0 and j==0:
+                        â = e(0)
+                        t.update()
+                        continue
+                    th = np.linspace(-math.pi/math.pow(2, i+1), math.pi/math.pow(2, i+1), n_g)
+                    cands = np.outer(â, np.cos(th)) + np.outer(e(j), np.sin(th))
+                    cands = cands / np.linalg.norm(cands, ord=2, axis=0)
+                    scores = S(np.dot(Xt, cands))
+                    best = np.argmax(scores)
+                    â = cands[:,best]
+                    t.postfix[1]["value"] = scores[best]
+                    t.update()
+        A[:,k] = â
+        Xt = Xt - np.outer(np.dot(Xt, â), â)
+
+        result = A[rev_order,0:(k+1)]
+        if sufficient and sufficient(result):
+            break
+
+    return result
+
+
+def projection(basis):
+    return np.linalg.solve(np.matmul(basis.transpose(), basis), basis.transpose())
```

Comment:
B^{t}BX = B^{t} ?  Is this giving the pseudo-inverse?

Reply:
Aha.  When you say it like that, I realize I should be using `np.linalg.pinv(basis)`.